[WIP] Add support of date from site config

Kdecherf opened the pull request:
This PR adds back the support of date to extract the publication date of articles. Partially based on #22. 

Fixes #67 

- [x] Support of `date` in ConfigBuilder
- [x] Support of date extraction in ContentExtractor
- [x] Support of date in Graby
- [x] Tests for ContentExtractor
- [x] Tests for Graby

Kdecherf commented:
Here are some refactoring in ContentExtractor. Tests are OK here.

But please note that i'm using `$this->{$entity}` to update properties from new methods. Are you okay with that?

j0k3r commented on src/Extractor/ContentExtractor.php:
```diff
@@ -669,4 +686,51 @@ private function extractBody($detectBody, $xpathExpression, \DOMNode $node, $typ
+        // we define the default callback here
+        if (!is_callable($returnCallback)) {
+            $returnCallback = function($e) {
+                return $e;
```
Looks like it'll be `trim()` most the time, I think we can use `trim()` as the default one.

Kdecherf replied:
Done

j0k3r commented on src/Extractor/ContentExtractor.php:
```diff
@@ -669,4 +686,51 @@ private function extractBody($detectBody, $xpathExpression, \DOMNode $node, $typ
+            return true;
+        } else {
+            return false;
+        }
```
This `else` isn't necessary since you already exit from the function in the below `if`

Kdecherf replied:
Done

j0k3r commented on src/Extractor/ContentExtractor.php:
```diff
@@ -669,4 +686,51 @@ private function extractBody($detectBody, $xpathExpression, \DOMNode $node, $typ
+    /**
+     * Extract and apply a callback to an entity according to a pattern
+     *
+     * @param string $entity            Entity to look for
```
I'm ok with the use of `$this->{$entity}` but please put some example for `$entity` in the phpDoc. This will be less obscure

Kdecherf replied:
Added more comment

nicosomb commented:
Here are some examples of html tags about dates: 

```
<meta name="pubdate" content="2015-11-26T07:11:02Z" >
<meta name='publishdate' content='201511261006'/>
<meta name="timestamp"  data-type="date" content="2015-11-25 22:40:25" />
<meta name="DC.date.issued" content="2015-11-26">
<meta property="article:published_time"  content="2015-11-25" />
<meta name="Date" content="2015-11-26" />
<meta property="bt:pubDate" content="2015-11-26T00:10:33+00:00">
<meta name="sailthru.date" content="2015-11-25T19:56:04+0000" />
<meta name="article.published" content="2015-11-26T11:53:00.000Z" />
<meta name="published-date" content="2015-11-26T11:53:00.000Z" />
<meta name="article.created" content="2015-11-26T11:53:00.000Z" />
<meta name="article_date_original" content="Thursday, November 26, 2015,  6:42 AM" />
<meta name="cXenseParse:recs:publishtime" content="2015-11-26T14:42Z"/>
<meta name="DATE_PUBLISHED" content="11/24/2015 01:05AM" />
<meta itemprop="datePublished" content="2015-11-26T11:53:00.000Z" />
<meta itemprop="datecreated" content="2015-11-26T11:53:00.000Z" />
<meta http-equiv="data" content="10:27:15 AM Thursday, November 26, 2015">
<time itemprop="datePublished" datetime="2015-10-16T03:56:15-05:00">
```

Does graby support all these tags?

Kdecherf commented:
Ah, thanks @nicosomb for examples.

The current implementation supports only time/@datetime out of the box (autodetect), I will try to add these in auto.

Another question: I don't like to return an integer for the datetime, it loses the original timezone. I would like to return a ISO8601-compliant string. Is it okay for you?
What graby must do if it fails to parse the date?

j0k3r commented:
I'm ok with that compliant date.
If parsing fail, log it and let the field empty

j0k3r commented:
@Kdecherf can you find some time to finish that PR? I would like to include it in the upcoming 1.6.0 🙂

Kdecherf commented:
@j0k3r sure, I will resume my work on it on next week. I hope to finish it before Feb, 13.

j0k3r commented:
This will be awesome 👍

Kdecherf commented:
Still WIP

j0k3r commented on src/Extractor/ContentExtractor.php:
```diff
@@ -731,7 +763,6 @@ private function extractBody($detectBody, $xpathExpression, \DOMNode $node, $typ
     }
 
     /**
-<<<<<<< HEAD
```
🤔

Kdecherf replied:
Well... I guess I won the right to rebase again :man_facepalming:

j0k3r commented:
So, I've re-added some tests from the previous PR.
I've also included the date in Graby.
I don't know if Graby should try to find the best date or this value should be in a config file instead.

Date is return as integer (from `strtotime`), if you want to convert it to ISO8601, feel free to update the PR.

All in all, it looks good to me.
I would like to merge it asap to avoid too much conflicts in the upcoming changes (mostly from cs-fixer fine tuning).

Kdecherf commented:
Thanks, I will resume the work on this PR tomorrow evening.

Sorry for the delay.

Kdecherf commented:
@j0k3r I replaced strtotime() with quick ISO8601 parsing and added two tests.

Currently there is only one tag supported by default for the extraction: `//time[@pubdate or @pubDate]`

j0k3r commented:
It's fine for me since it's the default behavior that was previously here.
I'll merge it and fix few things later.

Thanks again for the refacto! :call_me_hand:

nicosomb commented:
🤙 😁
